Add iam role

pccowboy opened the pull request:
Add a field to compute image to store an IAM role.  This role can be used to control AWS API access from within the image via a defined policy document.

bgupta committed:
Added header to foremancli calls to call API version 1

Commit:
fixes #1550 : Test existance before removing index and column.

In case where reference_id exist, but domain_id, hostgroup_id, domain_id don't, migration should'nt fail

Commit:
add an IAM role field to a compute image, to allow use of IAM to control access to AWS services within an instance.

Commit:
add an IAM role field to a compute image, to allow use of IAM to control access to AWS services within an instance.

Commit:
Merge branch 'add_ima_role' of https://github.com/pccowboy/foreman into add_ima_role

Commit:
Merge branch 'add_ima_role' of https://github.com/pccowboy/foreman into add_ima_role

Commit:
Merge branch 'add_ima_role' of https://github.com/pccowboy/foreman into add_ima_role

ohadlevy commented on app/models/orchestration/compute.rb:
```diff
@@ -141,5 +145,10 @@ def compute_update_required?
     end
 
+    def validate_compute_provisioning
+      image_uuid = compute_attributes[:image_id] || compute_attributes[:image_ref]
```
I'm not sure I understand what is the purpose of this validation? dont we set it up already some where else?

pccowboy replied:
the validation in ssh provisioning is what pulls in the image_uuid variable, so I duplicated that here.  I was uncertain where else I could add it.  The reason it needs to happen in both is that a compute resource  can be provisioned without needing ssh to bring it live.

I don't want to make the IAM role available as a fact, as we consider that a security issue, since facts are available to any user of an instance.  So, rather than pass it in the YAML, I just pull it in from the image attributes.

ohadlevy commented on app/models/orchestration/compute.rb:
```diff
@@ -56,7 +56,11 @@ def queue_compute_destroy
       logger.info "Adding Compute instance for #{name}"
-      self.vm = compute_resource.create_vm compute_attributes.merge(:name => name)
+      if (self.image.iam_role != "")
+        self.vm = compute_resource.create_vm compute_attributes.merge(:name => name, :iam_instance_profile_name => self.image.iam_role)
```
i think this is best to handle at the lib/model/foreman/ec2 class rathen then here, after all, it gets passed as part of the hash

pccowboy replied:
I did not see iam_instance_profile_name in the hash previously in debug output.

I am uncertain what fog and AWS will do if they are passed an empty iam_instance_profile_name, so I put the check here to make sure a blank one is never passed.  I also did not see anywhere obvious to add it in the lib/model/foreman/ec2 class.  If you have further feedback on how to add it there I'll be glad to add it.

ohadlevy commented on db/migrate/20120921000313_add_iam_role_to_images.rb:
```diff
@@ -0,0 +1,9 @@
+class AddIamRoleToImages < ActiveRecord::Migration
+  def self.up
+    add_column :images, :iam_role, :string
```
maybe I"m not 100% clear on what is the usage, but this is just to provide a default per image?

pccowboy replied:
Adding the iam role to the image seemed the easiest way to store one for now.  Later, I would like to put a dropdown into the virtual machine attributes page that lists the roles in a similar manner to the regions available, but I did not see an easy way to list the IAM roles with Fog.

abenari committed:
fixes #1028 - show and override puppet class parameters in the host form.

Commit:
fixes #1872 - Adds Puppet v3 support

pccowboy commented:
ping.

abenari committed:
fixes #1901

bgupta committed:
Merge pull request #187 from skottler/markitdown

Made the readme markdown and some minor cleanups

ohadlevy committed:
add subnets to safemode rendering

abenari committed:
fix minor css classes declaration.

abenari committed:
add pagination to compute resource

Commit:
Rails cache returns incorrect value for false values.

while this is fixed in Rails 3.1.4, a work around is required for now:

Settings getter method correctly returns false instead of nil.

ohadlevy committed:
fixes #1907 - audit log should contain puppet class assignments within a hostgroup

Commit:
minor API v1 fixes

* fix unknown variable name
* enable apipie params validation
* update config_template api doc

abenari committed:
fixes #1870 group compute resource by type in the new host form.

abenari committed:
fixes #1795 - When deleting compute resource, confirmation box should remind which compute resource is being deleted

abenari committed:
fixes #1772 - sort not working for facts.

Commit:
typo on index.json.rabl for environments.

added missing s char

ohadlevy committed:
fixes #1909 - add host audit entries to host show page

lzap committed:
redirecting foreman-config log output

Commit:
fixes #1913 Solaris architecture fact should always be hardwareisa

ohadlevy committed:
allows accessing token value from safe mode rendering

also added to_s method, just in case someone would do <%= token %>

Commit:
used media/show partial and added to tests

Commit:
added dns_id, dhcp_id, tft_ip attributes even though they are nested as children to be consistent with current api

Commit:
This commit adds most of the functionality required for API v1

The overall goal was to extract the existing JSON response
overall controllers, and to move them to a seperate name space.

* documentation was added to all requests (available under /apidoc)
this include same request and response generated using
    APIPIE_RECORD=examples rake test:api
* Also Fixes #1800, fixes #1419, fixes #981 and fixes #1774

whats missing:

1. Permissions checks for all new controllers
2. Nested routes (e.g. /hosts/id/reports)
3. running tests with oauth/basic_auth (instead of as_admin {get /})

Commit:
Removed the travis configuration

Commit:
Fixes validation for hostnames containing hyphens for NFS mounts

Commit:
fixes #1917 Solaris Sparc provisioning templates fix

GregSutcliffe committed:
Fixes #983 - implement 3 state boot

rbirnie committed:
fixes #1862 resets cookie after deleting hosts

rbirnie committed:
fixes #1918 dis-associate template kind from OS

GregSutcliffe committed:
Fixes #1920 - Template URLs should not be based on the browser address bar

rbirnie committed:
feature #1906 add Trend information

Signed-off-by: Amos Benari <abenari@redhat.com>
Signed-off-by: Ohad Levy <ohadlevy@gmail.com>

cyberkov committed:
fixed typo - "replay_address" should be called "reply_address"

added migration for typo in replay_adress variable

abenari committed:
fixed active tab on cached topbar.

abenari committed:
fixed hat location on login screen.

ohadlevy committed:
refs #1906 - checked in wrong migration file

GregSutcliffe committed:
Refs #1920 : Fix failing tests for Setting[:foreman_url] Ensure plain 'http' protocol is used in templates Add a test to ensure http is used when https is in @request

rbirnie committed:
fix #1922 add scrolling to graphs

Commit:
Fixed array structure in API v1 user controller test

rbirnie committed:
fix #1925 single value trends pages

rbirnie committed:
fix #1925 single value trends pages

Commit:
Added rabl gem version requirement

abenari committed:
Add organization and location to foreman.

This feature allows foreman to provide multi location, multi tenant and
multi organizations capablities.

the idea is that resources within foreman (e.g. hosts, subnets, users,
environments etc) can belong to one or more locations and organization,
effectivily hidding resources from users.

Organization may contain multiple locations, and Locations can belong
to multiple Organization, this is an extermily flexiable design,
however, it is up to the user to allocate the resources across the
organizations and locations.

When creating new hosts, the resources that can be consumed are only
resources in which exists in both the currently used organization and location.

fixes #1578
fixes #1593

xsuchy committed:
spec directory is obsolete

this live for some time in foreman-rpms.git repo

xsuchy committed:
debian directory is obsolete

this live for some time in foreman-rpms.git repo

Commit:
Improved spacing and active selection for dropdowns in the topbar

Commit:
fixes #1876 fix display of instance name and dns for AWS Use instance number when instances name is not set. Use private DNS entry when public dns is not available

Commit:
fixes #1934 fix display for AWS instances index Set proper status and remove start button for terminated instance

Commit:
Fixes #1911 - removes limit from change_column's in a few migrations

rbirnie committed:
Bug #1936 Allow user creation without email

ohadlevy committed:
refactor AuthSourceLdap Class

cleanup the classes, as it was aweful ruby.
TODO: improve the logic between existing user and new users on the fly creaetion

Commit:
redhat_register snippet: make RHN server configurable, support RHN hosted

Adds "spacewalk_type" and "spacewalk_host" parameters.

Commit:
Pin mocha to versions below 0.13.0

mbacovsky committed:
Fixed validation error handling and auth durin admin user creation

ohadlevy committed:
Revert "fix #1925 single value trends pages"

This reverts commit bf89a7f3e86c08eb1cd38ae473a0edc754f8131e.

this was merged by mistake

iNecas committed:
fixes #1937 - Foreman doesn't work well when run on subpath

* request_url is not enough to pass form controller to tftp
  Other variables (like script_name) need to be used from
  controller.url_options to properly generate the path in the PXE template.
* use rails helper to get the path to noVNC source
  Hardcoded value makes problems when running Foreman from subpath
* helper javascript method for generating path to foreman

abenari committed:
if the ovirt host uses spice now you can connect

Commit:
Fixes #1940 - Update all the unattended templates to use puppet agent

Commit:
fixes #1470 hosts with no report should not be negative

ohadlevy committed:
fixes #1589 provisioning vmware support

abenari committed:
fixes #1400 remove decimal point in distribution bar chart in dashboard.

abenari committed:
fixes #1143 Searching for hosts with parameters is broken.

abenari committed:
fixes #1942 Jarring on mouseover colour change in vm list

abenari committed:
fixes #1074 run distribution graph too vague

rbirnie committed:
update global parameters title to match other views

rbirnie committed:
fix trend labels for ForemanTrends

bgupta committed:
Removed deprecated settings for settings.yaml.example

ohadlevy committed:
fixes #1957 - remove old libvirt hypervisors code

Since foreman 1.0 libvirt hypervisors are now part of compute resources
this patch removes the old code and migrate any existing hypervisors to
be libvirt based compute resources.

GregSutcliffe committed:
Fixes #1823 - Use Compute Resources permissions for Virtual Machine tab

GregSutcliffe committed:
Show all Compute Resources if Compute Resource user filtering is not in use

GregSutcliffe committed:
Revert "Show all Compute Resources if Compute Resource user filtering is not in use"

This reverts commit 6bc745327d0e354c695eaef61162e54b9b198780.

abenari committed:
fixes #1637 adds spice support

added vm name and cursor release instruction to the spice console title.
this is a follow up to commit:057d497

ohadlevy committed:
fixes #1962 incorrect value returned when on the fly ldap is disabled

ohadlevy committed:
fixed wrong old hypervisor migration refs #1957

ohadlevy committed:
fixes wrong method name, #refs 1957

ohadlevy committed:
adds foreman versions to response headers

oourfali committed:
fixes #1961 Adding JSONP support

When using cross-domain requests, it is impossible to read the response
data, and process it.
Using JSONP allows a client, from a specific domain (for example
/domain1), to query data in another doamin (/domain2), and analyze it,
using a callback.

Such a request would be followed by "callback=" parameter. For example:
http://foreman-server:3000/statistics?format=json&callback=analyzeStats

And the response would be:
analyzeStats({"statistics":{"mem_totfree":0.0,"arch_count":{},
"env_count":{},"cpu_count":{},"mem_size":0,"swap_size":0,
"model_count":{},"mem_free":0,"swap_free":0,"klass_count":{},
"os_count":{},"mem_totsize":0.0}})

which will call the analyzeStats callback.
Whereas a regular request,
http://foreman-server:3000/statistics?format=json, will just contain the
JSON data.

More reading on JSONP:
JSONP - http://en.wikipedia.org/wiki/JSONP
JSONP Ruby impl -
www.simb.net/2012/02/06/ruby-and-jsonp
https://github.com/crohr/rack-jsonp/

ohadlevy committed:
do not report internal state over api

Commit:
fixes #1915 - initialise Puppet in master context

Explicitly specify the run_mode as being 'master' to ensure the config is
parsed in a puppetmaster context.

Commit:
smart proxies api - filtering by proxy type

Commit:
subnets api - removed required constraints from fields in update

The required constraints were blocking partial updates.

Commit:
set apiadmin user in headers before api controller tests

GregSutcliffe committed:
Fixes #1968 - Fix require paths for tests with new config/settings.rb

oourfali committed:
Fixed 1972: Supporting search hosts by uuid

This patch adds support for searching hosts by uuid.
It is useful when you have the uuid of a virtual machine, and want to
get the Foreman host properties.

Commit:
fixes #1886 adds Rackspace V2 support

Based on GregSutcliffe's initial pull request.

Commit:
fixes #1899 require HTTPS URL for oVirt API, now HTTPS-only

Commit:
removed params required => from update action

Commit:
Fixes bug in Puppetclass search_for by host

Commit:
fixes #1976 - fix path to Puppet to /usr/bin

ohadlevy committed:
Ensure existing org user search is unscoped

refs commit:611f5bff

ohadlevy committed:
ensure that internal users can login correctly when orgs is enabled

ohadlevy committed:
ensure no duplicate hosts can be created

ohadlevy committed:
make sure that default join objects of taxonomy are RW

ohadlevy committed:
Minor optimization

Since we are looking for the ID and the join table includes the ID
there is no reason to actaully fetch the data table it self.

ohadlevy committed:
oops, fixed typos

ohadlevy committed:
ensures that admin user can't be duplicated when running within an org

abenari committed:
Taxonomy (Orgs / Locations related fixes)

* fixed edit habtm with taxonomy join.
* code cleanup
* use sub query instead of inner join, it solved all the org problems :) and perform better.

iNecas committed:
Clear the thread values outside of handling request

Adding an around filter to clear the thread values. Without this there
is a risk that the thread value from previous request will be used in
other request, which can lead to security issues.

We clear the current user at the beginning of the request (except the
test environment, where it's being used to simulate user being logged
in). In the production, this should never happen, but it's better safe
than sorry.

iNecas committed:
Ignore value of User.current when authorizing

One exception is test environment, where the User.current is used to
simulate user being logged in.

In other environments, it's a security issue, in combination of not
clearing the thread values, it's possible to use the API without
specifying any credentials, provided the same thread handled some
other user before the attackers request.

iNecas committed:
Fix inter-test dependencies

After the thread clean-up, some tests were failing because they
implicitly expected User.current to be set. Stating the dependencies
explicitly.

dmitri-d committed:
added template_kinds controller to api

ohadlevy committed:
fog 1.8 has been released - updating gem file

ohadlevy committed:
ensure all taxonomies migration pass on pgsql as well.

ohadlevy committed:
Initial migration fixes for postgresql

ohadlevy committed:
Allow admin users to join Orgs

ohadlevy committed:
remove hypervisor vm info from tests refs #1957

ohadlevy committed:
fixes incorrect javascript at compute resources after taxonomy changes

abenari committed:
minore fix for the edit habtm.

abenari committed:
filtered users include admins, except in taxonomy form where it is no longer relevant to select admin users.

abenari committed:
admins can always reset the taxonomy filters to any

ohadlevy committed:
SQL improvments for paramaters in host edit section refs #2003

abenari committed:
replaced default taxonomy migration with a multiple actions for assigning hosts to orgs and locations.

ohadlevy committed:
fixes #2003 - Improve Parameterized classes values retrieval time

This patch adds a new classification class which
tries to fetch all values at once

The initial implementation cared about a name / value at a time
however getting all at once (e.g. during a puppet ENC call) is much more
common.

This patch creates in memory tree of all possible values and process
them in memory instead of asking the database for each possible match.

iNecas committed:
Fix the path for create medium API

ohadlevy committed:
fixes #2004 - Location and Orgs are visible in the host show page even when its disable

dmitri-d committed:
support for passing template_kind hash as input parameter during create and update operations

ohadlevy committed:
fixes #2010 - selecting an org/location does not expire topbar cache

ohadlevy committed:
refs #2010, fixes cache issues when selecting an org/location

ohadlevy committed:
fixes #2011 - Param Class optimization fails on Postges

Commit:
fixes #1994 - set vardir for puppet 3, use --configprint

When running Puppet 3 as a non-root user (i.e. foreman), both --confdir and
--vardir have to be specified as per Puppet #16637, else SSL settings dependent
on vardir will fail.

This adds a new puppetvardir setting used with puppetconfdir, and supports
puppetconfdir being a file or directory due to historic settings.yaml examples.

It now shells out and uses `puppet master --configprint` to retrieve settings
to avoid changes in Puppet settings initialization logic across versions (e.g.
run_mode on Puppet 3).

ohadlevy committed:
1.1RC1 released!

Commit:
fix merge conflict from rackspace section of template.

sodabrew commented:
Hi! Are most of these commits part of your PR, or are they merges? Would recommend doing a rebase of your branch up to master, rather than merging master into your branch.

pccowboy commented:
I have no idea why those commits are there, they are not part of my original PR, nor are they due to merges from me.  I _always_ rebase my branch to master before I send a PR.

I have quit trying to get my PRs into Foreman, as I have so far been unable to get anything accepted, with the exception of a couple of minor bugfixes.  It is a shame, as we are essentially maintaining our own fork, and when I leave this job (1-2 weeks from now) they will not have anyone on hand to pull in new functionality.  However, since the time to get a PR evaluated has stretched to months, we have had no choice.

FWIW, this PR in its original form is running fine on PopSugar's entire AWS production infrastructure.  We are appraoching nearly 100 instance, with plans for many more.  I am very curious why it takes so long to get a PR evaluated, given we have a ready test platform of that size, and up until a couple months ago, a dedicated person who wanted to contribute to foreman.

isratrade commented:
@pccowboy, I merged, rebased, and refactored your pull request for @ohadlevy to review
I moved some of the EC2 specific code from compute.rb to EC2.rb
if @ohadlevy accepts this, he will change the author to you.

ohadlevy commented:
closing in favor of #409 - thanks!

sodabrew commented:
@pccowboy -- PopSugar, small world! My wife used to work in finance there. [shameless plug] If you're looking for a new gig, BrightRoll is hiring: http://brightroll.com/careers/

pccowboy commented:
@aaron stone - I'll double your small world, and raise you one Katie 
Bishop - She and I have been talking this past week.  I'll have to 
connect up with you off-list, and get the skinny.

Thanks!

> Aaron Stone mailto:notifications@github.com
> Tuesday, February 12, 2013 8:47 AM
> 
> @pccowboy https://github.com/pccowboy -- PopSugar, small world! My 
> wife used to work in finance there. [shameless plug] If you're looking 
> for a new gig, BrightRoll is hiring: http://brightroll.com/careers/
> 
> ---
> 
> Reply to this email directly or view it on GitHub 
> https://github.com/theforeman/foreman/pull/160#issuecomment-13442133.
